test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+  | { name: 'wait'; ms: number }
+  | { name: 'waitForEvent'; event: string; count: number }
+  | { name: 'checkOut'; thread: string; label: string }
+  | { name: 'checkIn'; connection: string }
+  | { name: 'clear' | 'close' | 'ready' };
+
+type cmapPoolOptions = {
+  maxPoolSize?: number;
+  minPoolSize?: number;
+  maxIdleTimeMS?: number;
+  waitQueueTimeoutMS?: number;
+};
+
+type cmapEvent = {
+  type: string;
+  address?: 42;
+  connectionId?: number;
+  options?: 42 | cmapPoolOptions;
+  reason: string;
+};
+
+const knownTestKeys = [
+  'name',
+  'version',
+  'style',
+  'description',
+  'poolOptions',
+  'operations',
+  'error',
+  'events',
+  'ignore'
+];
+export type cmapTest = {
+  name?: string; // filename path added by the spec loader
+  version: number;
+  style: 'unit';
+  description: string;
+  poolOptions?: cmapPoolOptions;
+  operations: cmapOperation[];
+  error?: {
+    type: string;
+    message: string;
+    address?: number;
+  };
+  events?: cmapEvent[];
+  ignore?: string[];
+};
+
+const ALL_POOL_EVENTS = new Set([
+  ConnectionPool.CONNECTION_POOL_CREATED,
+  ConnectionPool.CONNECTION_POOL_CLOSED,
+  ConnectionPool.CONNECTION_POOL_CLEARED,
+  ConnectionPool.CONNECTION_CREATED,
+  ConnectionPool.CONNECTION_READY,
+  ConnectionPool.CONNECTION_CLOSED,
+  ConnectionPool.CONNECTION_CHECK_OUT_STARTED,
+  ConnectionPool.CONNECTION_CHECK_OUT_FAILED,
+  ConnectionPool.CONNECTION_CHECKED_OUT,
+  ConnectionPool.CONNECTION_CHECKED_IN
+]);
+
+function getEventType(event) {
+  const eventName = event.constructor.name;
+  return eventName.substring(0, eventName.lastIndexOf('Event'));
+}
+
+/**
+ * In the cmap spec and runner definition,
+ * a "thread" refers to a concurrent execution context
+ */
+class Thread {
+  _promise: Promise<void>;
+  _error: Error;
+  _killed = false;
+
+  _knownCommands: any;
+
+  // concurrent execution context
+  constructor(operations) {
+    this._promise = new Promise(resolve => {
+      this.start = () => resolve();
+    });
+
+    this._knownCommands = operations;
+  }
+
+  start: () => void;
+
+  queue(op: cmapOperation) {
+    if (this._killed || this._error) {
+      return;
+    }
+
+    this._promise = this._promise.then(() => this._runOperation(op)).catch(e => (this._error = e));
+  }
+
+  async _runOperation(op: cmapOperation) {
+    const operationFn = this._knownCommands[op.name];
+    if (!operationFn) {
+      throw new Error(`Invalid command ${op.name}`);
+    }
+
+    await operationFn(op);
+    await sleep();
+  }
+
+  async finish() {
+    this._killed = true;
+    await this._promise;
+    if (this._error) {
+      throw this._error;
+    }
+  }
+}
+
+/**
+ * Implements the spec test match function, see:
+ * [CMAP Spec Test README](https://github.com/mongodb/specifications/tree/master/source/connection-monitoring-and-pooling/tests#spec-test-match-function)
+ */
+const compareInputToSpec = (input, expected) => {
+  if (expected === 42) {
```

Comment:
The spec says `If expected is "42" or 42:` so I think we should be checking the string case as well, even if there are none currently.

Reply:
that was intentional actually - the thought was we keep it as is and update it whenever we sync such a test, because it's going to fail at that point, and it reduces the mental load of trying to check everything the function does to just the things that are relevant